usdoj-crt/beta-ada
# Technical Assistance Landing Page Components

This is a BIG pull request. 

There's lots of new JS code that needs to be reviewed in depth. This PR up for A11y testing and review specifically. 
A11y review passed 6/2/2022 - no concerns.

Moving onto code review.

### To test: 
1. Pull down the branch locally
4. Run 'npm run test:ui'
5. All UI tests should pass
6. Navigate to localhost:4000/resources
7. Fiddle with the controls in the left hand nav. 
8. Confirm that the badges remove the appropriate checkboxes when clicked
9. Confirm the checkboxes add the appropriate badges when clicked
10. Confirm the page content is filtered when checkboxes are clicked. 
11. The filtering is additive. Meaning that it adds items that contain a certain tag. So more things should be returned when more filters are added. 
12. Confirm that the url is being altered when adding filters via the checkboxes. 
13. Try copying the link and heading to a new browser tab and pasting it in. The new page should populate with all of the filters in the correct state. 

### Changes:
 Ignore the compiled JS assets in the /dist folder. 
1. main.js - added new class .noAnchor to hide certain h2s from AnchorJS
14. taSelectors.js - main script handling the functionality of the checkboxes on the resources landing page. 
15. badgeLogic.js - attaches an onclick listener to the badges that updates the state of the url and dom when they are removed.
16. badgeTemplate.js - generates the markup for each badge 
17. renderBadges.js - loops over the current state of the page and creates the badges using the badgeTemplate and adds the onClick listener from badgeLogic to each badge.
18. pageNumber.js - refers to the refactored getSearchParam function
19. paginationButtontemplate.js - refers to the refactored getSearchParam function
20. totalResultsTemplate.js - refactored to supply new text in different cases depending on the number of results. Also takes second argument to easily update text.
21. applyFocusStyling.js - refers to the refactored getSearchParam function
22. checkURLOnLoad.js - splits the search param array into a string
23. constants.js - includes list of all available tags now
24. expandTarget.js - expands or collapses the category drop down button
25. findFocus.js - returns the element that is currently in focus 
26. moveFocus.js - focuses the next or previous item in the dom depending on the case
27. renderSearchPage.js - refers to refactored totalResults function
28. replaceHistory.js - updates/replaces the history of the browser
29. searchParamUtils.js - Used to be offsetUtils.js, refactored to include the option to pass in param strings
30. toggleVisibility.js - toggles the visibility of specific items in the article list
31. updateTaSelectors.js - updates the state of the checkboxes

### Additional Tests:
1. Manual a11y testing with screen readers like Voice Over, JAWS and NVDA. - Done.
32. Automated testing via Pa11y - Done.
33. UI testing via Playwright. These have been written, and all pass. Need to confirm during JS review session.
34. Browser-side testing via Axe Devtools, Waves and Lighthouse.  - Done
35. Potentially user testing.
